src/inventory/transaction/EnchantTransaction.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\inventory\transaction;
+
+use pocketmine\event\inventory\EnchantItemEvent;
+use pocketmine\item\enchantment\EnchantmentHelper;
+use pocketmine\item\enchantment\EnchantmentOption;
+use pocketmine\item\Item;
+use pocketmine\item\ItemTypeIds;
+use pocketmine\player\Player;
+use function count;
+
+class EnchantTransaction extends InventoryTransaction{
+
+	private Item $inputItem;
+	private Item $outputItem;
+
+	public function __construct(
+		Player $source,
+		private readonly EnchantmentOption $option,
+		private readonly int $cost
+	){
+		parent::__construct($source);
+	}
+
+	private function validateOutput() : void{
+		$enchantedInput = EnchantmentHelper::enchantItem($this->inputItem, $this->option->getEnchantments());
+
+		if(!$this->outputItem->equalsExact($enchantedInput) ||
+			$this->outputItem->getEnchantments() !== $enchantedInput->getEnchantments()
+		){
+			throw new TransactionValidationException("Invalid output item");
+		}
+	}
+
+	private function validateFiniteResources(int $lapisSpent) : void{
+		if($lapisSpent !== $this->cost){
+			throw new TransactionValidationException("Expected the amount of lapis lazuli spent to be $this->cost, but received $lapisSpent");
+		}
+
+		$xpLevel = $this->source->getXpManager()->getXpLevel();
+		$requiredXpLevel = $this->option->getRequiredXpLevel();
+
+		if($xpLevel < $requiredXpLevel){
+			throw new TransactionValidationException("Player's XP level $xpLevel is less than the required XP level $requiredXpLevel");
+		}
+		if($xpLevel < $this->cost){
+			throw new TransactionValidationException("Player's XP level $xpLevel is less than the XP level cost $this->cost");
+		}
+	}
+
+	public function validate() : void{
+		if(count($this->actions) < 1){
+			throw new TransactionValidationException("Transaction must have at least one action to be executable");
+		}
+
+		/** @var Item[] $inputs */
+		$inputs = [];
+		/** @var Item[] $outputs */
+		$outputs = [];
+		$this->matchItems($outputs, $inputs);
+
+		$lapisSpent = 0;
+		foreach($inputs as $input){
+			if($input->getTypeId() === ItemTypeIds::LAPIS_LAZULI){
+				$lapisSpent = $input->getCount();
+			}else{
+				if(isset($this->inputItem)){
```

Comment:
For static analyzability, it's better to make these fields `null` rather than putting them in the `uninitialized` state.

However, this will likely generate some PHPStan errors in `callExecuteEvent`. Use conditions with `AssumptionFailedError` for that.